subsys/sd/sdio.c

```diff
@@ -56,23 +63,997 @@ int sdio_send_ocr(struct sd_card *card, uint32_t ocr)
+			break;
+		}
+		/* Check to see if read tuple matches any we should look for */
+		for (int i = 0; i < tuple_count; i++) {
+			if (tpl_code == tuples[i]) {
+				match_tpl = true;
+				break;
+			}
+		}
+		if (match_tpl) {
+			/* tuple chains may be maximum of 255 bytes long */
+			memset(data, 0, 255);
+			for (int i = 0; i < tpl_link; i++) {
+				ret = sdio_io_rw_direct(func->card, SDIO_IO_READ,
+					func->num, cis_ptr++, 0, data + i);
+				if (ret) {
+					return ret;
+				}
+			}
+			num++;
+			match_tpl = false;
+			/* Decode the CIS data we read */
+			sdio_decode_cis(&func->cis, func->num, data,
+				tpl_code, tpl_link);
+		} else {
+			/* Advance CIS pointer */
+			cis_ptr += tpl_link;
+		}
+	} while (num < tuple_count);
+	LOG_DBG("SDIO CIS max block size for func %d: %d", func->num,
+		func->cis.max_blk_size);
+	return ret;
+}
+
+static int sdio_set_bus_width(struct sd_card *card, enum sdhc_bus_width width)
+{
+	uint8_t reg_bus_interface = 0U;
+	int ret;
+
+	ret = sdio_io_rw_direct(card, SDIO_IO_READ, SDIO_FUNC_NUM_0,
+		SDIO_CCCR_BUS_IF, 0, &reg_bus_interface);
+	if (ret) {
+		return ret;
+	}
+	reg_bus_interface &= ~SDIO_CCCR_BUS_IF_WIDTH_MASK;
+	switch (width) {
+	case SDHC_BUS_WIDTH1BIT:
+		reg_bus_interface |= SDIO_CCCR_BUS_IF_WIDTH_1_BIT;
+		break;
+	case SDHC_BUS_WIDTH4BIT:
+		reg_bus_interface |= SDIO_CCCR_BUS_IF_WIDTH_4_BIT;
+		break;
+	case SDHC_BUS_WIDTH8BIT:
+		reg_bus_interface |= SDIO_CCCR_BUS_IF_WIDTH_8_BIT;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+	ret = sdio_io_rw_direct(card, SDIO_IO_WRITE, SDIO_FUNC_NUM_0,
+		SDIO_CCCR_BUS_IF, reg_bus_interface, &reg_bus_interface);
+	if (ret) {
+		return ret;
+	}
+	/* Card now has changed bus width. Change host bus width */
+	card->bus_io.bus_width = width;
+	ret = sdhc_set_io(card->sdhc, &card->bus_io);
+	if (ret) {
+		LOG_DBG("Could not change host bus width");
+	}
+	return ret;
+}
+
+static inline void sdio_select_bus_speed(struct sd_card *card)
+{
+	if (card->host_props.host_caps.sdr104_support &&
+		(card->cccr_flags & SDIO_SUPPORT_SDR104) &&
+		(card->host_props.f_max >= SD_CLOCK_208MHZ)) {
+		card->card_speed = SD_TIMING_SDR104;
+	} else if (card->host_props.host_caps.ddr50_support &&
+		(card->cccr_flags & SDIO_SUPPORT_DDR50) &&
+		(card->host_props.f_max >= SD_CLOCK_50MHZ)) {
+		card->card_speed = SD_TIMING_DDR50;
+	} else if (card->host_props.host_caps.sdr50_support &&
+		(card->cccr_flags & SDIO_SUPPORT_SDR50) &&
+		(card->host_props.f_max >= SD_CLOCK_100MHZ)) {
+		card->card_speed = SD_TIMING_SDR50;
+	} else if (card->host_props.host_caps.high_spd_support &&
+		(card->switch_caps.bus_speed & SDIO_SUPPORT_HS) &&
+		(card->host_props.f_max >= SD_CLOCK_50MHZ)) {
+		card->card_speed = SD_TIMING_SDR25;
+	} else {
+		card->card_speed = SD_TIMING_SDR12;
 	}
 }
 
+/* Applies selected card bus speed to card and host */
+static int sdio_set_bus_speed(struct sd_card *card)
+{
+	int ret, retries = CONFIG_SD_RETRY_COUNT;
+	int timing = 0;
+	uint8_t target_speed = 0;
```

Comment:
I think these two don't need to be initialized.

Reply:
Yeah, you're right- removed both